MSAL/src/native_auth/controllers/sign_up/MSALNativeAuthSignUpController.swift

```diff
@@ -164,52 +164,63 @@ final class MSALNativeAuthSignUpController: MSALNativeAuthBaseController, MSALNa
             MSALLogger.log(level: .info, context: context, format: "Successful signup/start request")
             let challengeResult = await performAndValidateChallengeRequest(continuationToken: continuationToken, context: context)
             return handleSignUpChallengeResult(challengeResult, username: username, event: event, context: context)
-        case .attributeValidationFailed(let invalidAttributes):
+        case .attributeValidationFailed(let apiError, let invalidAttributes):
+            let context = updateContext(context, with: apiError.getHeaderCorrelationId())
             MSALLogger.log(
                 level: .error,
                 context: context,
                 format: "attribute_validation_failed received from signup/start request for attributes: \(invalidAttributes)"
             )
             let message = String(format: MSALNativeAuthErrorMessage.attributeValidationFailedSignUpStart, invalidAttributes.description)
-            let error = SignUpStartError(type: .generalError, message: message)
+            let error = apiError.toSignUpStartPublicError(context: context, message: message)
             return .init(.attributesInvalid(invalidAttributes), telemetryUpdate: { [weak self] result in
                 // The telemetry event always fails because the attribute validation failed
                 self?.stopTelemetryEvent(event, context: context, delegateDispatcherResult: result, controllerError: error)
             })
         case .redirect:
-            let error = SignUpStartError(type: .browserRequired)
+            let error = SignUpStartError(type: .browserRequired, correlationId: context.correlationId())
             stopTelemetryEvent(event, context: context, error: error)
             MSALLogger.log(level: .error,
                            context: context,
                            format: "Redirect error in signup/start request \(error.errorDescription ?? "No error description")")
             return .init(.error(error))
-        case .error(let apiError):
-            let error = apiError.toSignUpStartPublicError()
+        case .error(let apiError),
+             .unauthorizedClient(let apiError):
+            let context = updateContext(context, with: apiError.getHeaderCorrelationId())
+            let error = apiError.toSignUpStartPublicError(context: context)
             stopTelemetryEvent(event, context: context, error: error)
             MSALLogger.log(level: .error,
                            context: context,
                            format: "Error in signup/start request \(error.errorDescription ?? "No error description")")
             return .init(.error(error))
         case .invalidUsername(let apiError):
-            let error = SignUpStartError(type: .invalidUsername, message: apiError.errorDescription)
+            let context = updateContext(context, with: apiError.getHeaderCorrelationId())
+            let error = SignUpStartError(
```

Comment:
can you use the method `apiError.toSignUpStartPublicError()` in this case?

Reply:
No, because `apiError.toSignUpStartPublicError()` checks only for the `error: MSALNativeAuthSignUpStartOauth2ErrorCode` property. 

In this case, the property is `.invalidRequest`, and the type of public error is decided at the Validator level using the errorCodes.

This error handling across so many layers is also a good candidate to be simplified in the future.

Reply:
Yeah, it's not so straightforward. Would be ideal to have an error creation more homogeneous.